neps/nep-0518.md

```diff
@@ -0,0 +1,221 @@
+---
+NEP: 518
+Title: Web3-Compatible Wallets Support
+Authors: Aleksandr Shevchenko <alex.shevchenko@aurora.dev>, Michael Birch <michael.birch@aurora.dev>
+Status: New
+DiscussionsTo: https://github.com/near/NEPs/issues/518
+Type: Protocol
+Version: 1.0.0
+Created: 2023-11-15
+LastUpdated: 2024-07-22
+---
+
+## Summary
+
+This NEP describes the protocol changes needed to support the usage of Ethereum-compatible wallets (Web3 wallets), for example Metamask, on Near native applications. That is to say, with this protocol change all Metamask users can become Near users without installing any additional software; from their perspective Near will appear as just another network they can choose from (similar to Aurora today).
+
+This is accomplished through two key protocol changes:
+
+1. Ethereum-like addresses (i.e. account IDs of the form `^0x[a-f0-9]{40}$`) are implicit accounts on Near (i.e. can be created via a `Transfer` action). We call these "eth-implicit accounts".
+2. Unlike the current implicit accounts (64-character hex-encoded), eth-implicit accounts do not have any access keys added to them on creation. Instead, these accounts will have a special contract deployed to them automatically called the "wallet contract". This wallet contract enables the owner of the Ethereum address corresponding to the eth-implicit account ID to sign transactions with their Ethereum private key, thus providing similar functionality to the default access key of 64-character implicit accounts.
```

Comment:
Currently it is impossible to deploy a different contract to a new eth implicit account. This could be enabled if the wallet contract was upgraded to allow users to take the `DeployContract` action or if the wallet contract allowed adding full access keys. However, this is a dangerous feature because it could result in users no longer being able to control their account using Ethereum-like transactions.

I cannot think of good use cases where the benefits of being able to deploy a different contract are worth the risk of a user accidentally locking themselves out of their account. But if there is one in the future then we can consider upgrading the wallet contract in another protocol change.